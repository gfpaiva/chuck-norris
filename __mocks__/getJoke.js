module.exports.getJoke = {
  categories: [
    'science',
  ],
  created_at: '2016-05-01 10:51:41.584544',
  icon_url: 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
  id: 'rgxflbh9qx-i51rt3o3-xq',
  updated_at: '2016-05-01 10:51:41.584544',
  url: 'https://api.chucknorris.io/jokes/rgxflbh9qx-i51rt3o3-xq',
  value: 'Chuck Norris describes human beings as \'a sociable holder for blood and guts\'.'
};

module.exports.getOtherJoke = {
  categories: [
    'science',
  ],
  created_at: '2016-05-01 10:51:41.584544',
  icon_url: 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
  id: 'rgxflbh9qx-i51rt3o3-xq',
  updated_at: '2016-05-01 10:51:41.584544',
  url: 'https://api.chucknorris.io/jokes/rgxflbh9qx-i51rt3o3-xq',
  value: 'Nothing can escape the gravity of a black hole, except for Chuck Norris. Chuck Norris eats black holes. They taste like chicken',
};
