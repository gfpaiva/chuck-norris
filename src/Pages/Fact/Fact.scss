@import '../../Scss/root';

.fact {
  &__back,
  &__refresh {
    @include rfs($fs-medium);
    display: inline-block;
    transition: transform .5s ease;
  }
  &__back {
    &:hover { transform: translateX(-30%); }
  }

  &__refresh {
    &:hover { transform: rotate(-90deg) }
  }

  &__quote {
    @include rfs($fs-big);
    font-weight: 100;
    font-style: italic;
  }
}
