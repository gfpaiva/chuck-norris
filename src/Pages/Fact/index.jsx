import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import Spinner from 'react-spinkit';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Link } from 'react-router-dom';
import { FaArrowLeft, FaUndo } from 'react-icons/fa';

import FactsActions from '../../Store/facts/actions';

import Card from '../../Components/Card';
import CategoryGrid from '../../Components/CategoryGrid';

import { getCategory } from '../../Utils/Categories';

import './Fact.scss';

let detailCategory;

function Fact({ match: { params: { category } }, loadFactsRequest, facts }) {
  const { fact, loading, error } = facts;

  const handleRefresh = (e) => {
    e.preventDefault();
    loadFactsRequest(category);
  };

  useEffect(() => {
    loadFactsRequest(category);
    detailCategory = getCategory(category) ? getCategory(category) : { content: category, icon: '' };
    window.scrollTo(0, 0);
  }, [category]);

  return (
    <div className="fact">
      <Link className="fact__back" to="/">
        <FaArrowLeft />
      </Link>

      {detailCategory && (
        <p>
          <strong>
            {detailCategory.icon}
            &nbsp;
            {detailCategory.content}
          </strong>
          &nbsp;
          - Chuck Norris Fact:
        </p>
      )}

      {error && (
        <h2 className="message message--error text-center">
          Ocorreu um erro ao recuperar os dados
          <span role="img" aria-label="Disappointed Face">ðŸ˜ž</span>
        </h2>
      )}

      {loading && (
        <div className="text-center">
          <Spinner name="ball-pulse-sync" />
        </div>
      )}

      {!!fact && (
        <Card className={`fact__card animated faster ${loading ? 'zoomOut' : 'zoomIn'}`}>
          <div className="fact__card-content">

            <p className="fact__quote text-center">
              &quot;
              {fact}
              &quot;
            </p>

            <div className="text-right">
              <a
                className="fact__refresh"
                href="javascript"
                onClick={handleRefresh}
              >
                <FaUndo />
              </a>
            </div>
          </div>
        </Card>
      )}

      <div className="fact__categories">
        <p className="my-4">Navegue por outra categoria: </p>
        <CategoryGrid active={category} />
      </div>
    </div>
  );
}

Fact.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      category: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  loadFactsRequest: PropTypes.func.isRequired,
  facts: PropTypes.shape({
    fact: PropTypes.string,
    loading: PropTypes.bool,
    error: PropTypes.bool,
  }).isRequired,
};

const mapStateToProps = ({ facts }) => ({ facts });
const mapDispatchToProps = dispatch => bindActionCreators(FactsActions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Fact);
