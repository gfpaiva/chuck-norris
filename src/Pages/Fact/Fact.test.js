import React from 'react';
import wait from 'waait';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import store from '../../Store';

import Fact from './index';

import { getJoke, getOtherJoke } from '../../../__mocks__/getJoke';

window.scrollTo = jest.fn();

describe('<Fact />', () => {
  it('should mount properly', async () => {
    fetch
      .once(JSON.stringify(getJoke))

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{ pathname: '/category/dev' }]}>
          <Fact match={{ params: { category: 'dev' } }} />
        </MemoryRouter>
      </Provider>
    );

    // Loading state
    expect(wrapper.find('Fact')).toMatchSnapshot();

    await wait();
    wrapper.update();

    expect(wrapper.find('.card.fact__card').text()).toBe(`"${getJoke.value}"`);
  });

  it('should handle with refresh joke', async () => {
    fetch
      .once(JSON.stringify(getJoke))
      .once(JSON.stringify(getOtherJoke));

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{ pathname: '/category/dev' }]}>
          <Fact match={{ params: { category: 'dev' } }} />
        </MemoryRouter>
      </Provider>
    );

    await wait();
    wrapper.update();

    wrapper.find('.fact__refresh').simulate('click');

    await wait();
    wrapper.update();

    expect(wrapper.find('.card.fact__card').text()).toBe(`"${getOtherJoke.value}"`);
  });
});
