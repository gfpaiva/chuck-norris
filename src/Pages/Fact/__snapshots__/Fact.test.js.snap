// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fact /> should mount properly 1`] = `
<Fact
  facts={
    Object {
      "error": false,
      "fact": "",
      "loading": true,
    }
  }
  loadFactsFail={[Function]}
  loadFactsRequest={[Function]}
  loadFactsSuccess={[Function]}
  match={
    Object {
      "params": Object {
        "category": "dev",
      },
    }
  }
>
  <div
    className="fact"
  >
    <Link
      className="fact__back"
      to="/"
    >
      <a
        className="fact__back"
        href="/"
        onClick={[Function]}
      >
        <FaArrowLeft>
          <IconBase
            attr={
              Object {
                "viewBox": "0 0 448 512",
              }
            }
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
                key="0"
              />
            </svg>
          </IconBase>
        </FaArrowLeft>
      </a>
    </Link>
    <p>
      <strong>
        ðŸ¤“
        Â 
        DEV
      </strong>
      Â  - Chuck Norris Fact:
    </p>
    <div
      className="text-center"
    >
      <Spinner
        fadeIn="full"
        name="ball-pulse-sync"
        noFadeIn={false}
        overrideSpinnerClassName=""
      >
        <div
          className="sk-fade-in sk-spinner ball-pulse-sync"
        >
          <div
            key="0"
          />
          <div
            key="1"
          />
          <div
            key="2"
          />
        </div>
      </Spinner>
    </div>
    <div
      className="fact__categories"
    >
      <p
        className="my-4"
      >
        Navegue por outra categoria: 
      </p>
      <Connect(CategoryGrid)
        active="dev"
      >
        <CategoryGrid
          active="dev"
          categories={Array []}
          dispatch={[Function]}
        >
          <div
            className="category-grid"
          />
        </CategoryGrid>
      </Connect(CategoryGrid)>
    </div>
  </div>
</Fact>
`;
